### We have built text classification models using scikit learn, now let look inside the model to understand how it does prediction and what the most important features are
Fist let's quickly build the model again using SGDClassifier

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import SGDClassifier
from sklearn import metrics

In [2]:
df = pd.read_csv('bbc-text.csv')
print(df.shape, df['category'].nunique())
df.head(2)

(2225, 2) 5


category                                               text
0      tech  tv future in the hands of viewers with home th...
1  business  worldcom boss  left books alone  former worldc...

In [3]:
df['category'].value_counts()

sport            511
business         510
politics         417
tech             401
entertainment    386
Name: category, dtype: int64

In [4]:
sgd = Pipeline([("tfidf_vector_com",TfidfVectorizer()), ( "clf", SGDClassifier())])

In [5]:
%%time
X_train, X_test, y_train, y_test = train_test_split(
    df['text'], df['category'], test_size=.2, stratify=df['category'], random_state=42)
sgd.fit(X_train, y_train)
pred_test = sgd.predict(X_test)
pred_train = sgd.predict(X_train)
print("test accuracy", np.mean(pred_test == y_test))
print("train accuracy", np.mean(pred_train == y_train))
print(metrics.classification_report(y_test, pred_test))

test accuracy 0.9775280898876404
train accuracy 1.0
               precision    recall  f1-score   support

     business       0.97      0.97      0.97       102
entertainment       0.96      0.99      0.97        77
     politics       0.98      0.96      0.97        84
        sport       1.00      1.00      1.00       102
         tech       0.97      0.96      0.97        80

     accuracy                           0.98       445
    macro avg       0.98      0.98      0.98       445
 weighted avg       0.98      0.98      0.98       445

CPU times: user 1.92 s, sys: 274 ms, total: 2.2 s
Wall time: 899 ms


### Now we have the model created, let dissect the model to gain insights
The model pipeline conatains two steps: TfidfVectorizer for feature extraction, and SGDClassifier as classifier

In [6]:

sgd

Pipeline(steps=[('tfidf_vector_com', TfidfVectorizer()),
                ('clf', SGDClassifier())])

The tfidf (a common term weighting scheme in information retrieval) values that were fed into SGD classifier

In [7]:
print(sgd['tfidf_vector_com'].idf_.shape)
sgd['tfidf_vector_com'].idf_

(26795,)


array([6.69317081, 2.47121513, 7.7917831 , ..., 7.7917831 , 7.7917831 ,
       7.38631799])

Check the model target classes and the feature coefficients for the classes

In [8]:
print(sgd.classes_)
print(sgd['clf'].coef_.shape) # <-- (number of classes, number of features)
sgd['clf'].coef_

['business' 'entertainment' 'politics' 'sport' 'tech']
(5, 26795)


array([[-0.06424924,  0.32160756,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [-0.04559256, -0.08990268,  0.        , ...,  0.05277412,
        -0.10943533,  0.        ],
       [ 0.04959907,  0.49723019,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.02885362, -0.5829393 ,  0.        , ...,  0.        ,
         0.        ,  0.        ],
       [ 0.05720101,  0.01831583,  0.        , ...,  0.        ,
         0.        ,  0.        ]])

From above we can see there are total 5 classification classes and there are total 26795 features, those 26795 features correspond to 26795 TFIDF values encoded from the same number of word tokens, let find out the most import word tokens used to calculate the probability for the target class

For business class, it is in the first index position 0 in the classes_ attribute arracy, we can use the same index to get the feature coefficients for "business"

In [9]:
sgd['clf'].coef_[0]

array([-0.06424924,  0.32160756,  0.        , ...,  0.        ,
        0.        ,  0.        ])

The maximum coefficient index and the max coefficient value is

In [10]:
idx = np.argmax(sgd['clf'].coef_[0])
print(idx)
sgd['clf'].coef_[0][idx]

13260


2.4005051825497112

Lets find out the word token correspond to this maximum coefficient using the vocabulary attibute from TfidfVecterizer 

In [11]:
# Convert word to index sequence dictionary to index to word for easy lookup word by index
idx_to_word = {idx:word for (word, idx) in sgd['tfidf_vector_com'].vocabulary_.items()}
idx_to_word[idx]

'its'

Putting all these together. lest get top N features for a class

In [18]:
def top_n_features_by_coef(n, classname):
    """
        Args:
            n: the top number of words by coefficient,
            classname: the class label
        Returns:
            top or bottom n words with coefficients
    """
    class_idx = np.where(sgd.classes_== classname)[0][0]
    idx_coef = sorted(
        [(i,v) for (i, v) in enumerate(sgd['clf'].coef_[class_idx])], key=lambda e: e[1],reverse=True)
    top_n_idx_coef = idx_coef[:n]
    bottom_n_idx_coef = idx_coef[-n:] # top negative N words, sort asscending
    top_word_coef = list(map(lambda e: (idx_to_word[e[0]], round(e[1], 4)), top_n_idx_coef))
    bottom_word_coef = list(map(lambda e: (idx_to_word[e[0]], round(e[1], 4)), bottom_n_idx_coef))
    df = pd.DataFrame(top_word_coef, columns=[f"{classname}_word", 'coef'])
    df_bottom = pd.DataFrame(bottom_word_coef, columns=[f"{classname}_word", 'coef'])
    return pd.concat([df, df_bottom])

What are the word tookens by top positive and negative coeffients?

In [19]:
n = 10
df_list = []
for _class in sgd.classes_:
    df_list.append(top_n_features_by_coef(n, _class))
pd.concat(df_list, axis=1)

business_word    coef entertainment_word    coef politics_word    coef  \
0           its  2.4005               film  3.4698         party  2.3294   
1          bank  2.3939               show  2.7745         blair  2.3008   
2      economic  2.3341              music  2.3008    government  1.9766   
3        shares  2.1957             singer  2.1822        labour  1.9396   
4          firm  1.9372              album  2.1260          lord  1.9291   
5       company  1.8363               star  1.8786            mr  1.8367   
6      business  1.7527               band  1.8117     committee  1.8335   
7    investment  1.5838           festival  1.7344     secretary  1.8280   
8         sales  1.5324                 tv  1.6898      minister  1.8098   
9        market  1.4782              chart  1.6275         straw  1.7584   
0     committee -1.1054             labour -0.8851          club -1.0626   
1            uk -1.1491           minister -0.9839     companies -1.0649   
2        people -1.1704                 to -1.0174           net -1.1011   
3         brown -1.1706               firm -1.0257            tv -1.1104   
4         music -1.2938         technology -1.0404         rugby -1.1504   
5         straw -1.3013           internet -1.0578      software -1.1647   
6       britain -1.3039              games -1.2107            us -1.4286   
7          film -1.3973            players -1.3294          film -1.4291   
8         blair -1.4107             online -1.3732         music -1.4483   
9            he -1.5828               game -1.4735      economic -1.5548   

  sport_word    coef    tech_word    coef  
0        cup  1.7925     computer  2.4449  
1    players  1.7733   technology  2.3884  
2      match  1.7472       online  2.3405  
3       club  1.6202     software  2.3355  
4  athletics  1.6165         game  2.0802  
5  liverpool  1.6002        games  2.0757  
6      rugby  1.4954      digital  2.0665  
7      coach  1.4910        users  1.7263  
8        win  1.4361          ink  1.7228  
9   champion  1.3948     internet  1.6905  
0      album -1.0300  nominations -0.7940  
1         uk -1.0436   complaints -0.8121  
2     shares -1.1002        after -0.8312  
3     singer -1.1988         corp -0.8377  
4      music -1.2189     vodafone -0.8712  
5       said -1.2621        sales -0.8992  
6        its -1.2958      chernin -0.9470  
7       film -1.3940       shares -0.9972  
8         of -1.7923          ict -1.0351  
9         mr -2.9217   government -1.0401

From above we can see the word tokens with highest positive coefficients for "business" are "its", "bank", "economic", and "shares", etc. And the word token with highest negative coefficients for "business" are "committee", "uk", "people", "brown", and "music", etc. The word token with highest negative coefficients are more likely to be associated with high coefficients in some other categories, for example, "committee" appears in top coefficients for "politics" and "music" appears in top coefficients for "entertainment"

What what the top positive and negative word tokens by the feature output calculated from tfidf * coefficient

In [20]:
def top_n_features_by_feature_output(n, classname):
    """
        Args:
            n: the top number of words by coefficient,
            classname: the class label
        Returns:
            top or bottom n words with coefficients
    """
    class_idx = np.where(sgd.classes_== classname)[0][0]
    feature_output = sgd['tfidf_vector_com'].idf_ * sgd['clf'].coef_
    idx_coef = sorted(
        [(i,v) for (i, v) in enumerate(feature_output[class_idx])], key=lambda e: e[1],reverse=True)
    top_n_idx_coef = idx_coef[:n]
    bottom_n_idx_coef = idx_coef[-n:] # top negative N words, sort asscending
    top_word_coef = list(map(lambda e: (idx_to_word[e[0]], round(e[1], 4)), top_n_idx_coef))
    bottom_word_coef = list(map(lambda e: (idx_to_word[e[0]], round(e[1], 4)), bottom_n_idx_coef))
    df = pd.DataFrame(top_word_coef, columns=[f"{classname}_word", 'coef'])
    df_bottom = pd.DataFrame(bottom_word_coef, columns=[f"{classname}_word", 'coef'])
    return pd.concat([df, df_bottom])

In [15]:
n = 10
df_list = []
for _class in sgd.classes_:
    df_list.append(top_n_features_by_feature_output(n, _class))
pd.concat(df_list, axis=1)

business_word    coef entertainment_word     coef politics_word     coef  \
0     crossrail  9.4400               film  11.2000           ict  13.4400   
1           wto  8.8100             ballet  10.8200         straw   9.1900   
2          bank  8.7700            hendrix  10.0100          lord   8.1700   
3   datamonitor  8.2600              album   9.5600      councils   8.0500   
4      economic  8.1100            gallery   9.5100         blair   8.0100   
5        shares  8.1000             singer   9.0900       snooker   7.8400   
6        boeing  7.9300           festival   8.3600         duchy   7.7000   
7          feta  7.8800               show   8.2700         party   7.6300   
8         davos  7.4400           freeview   7.9300     committee   7.2900   
9       plastic  7.3300               band   7.7300          ukip   7.2700   
0         duchy -4.6465              savoy  -3.9825       ukraine  -4.6057   
1          spam -4.7505             madden  -4.0158          film  -4.6123   
2            g8 -4.8147            players  -4.0562         music  -4.6370   
3           ict -4.8678               game  -4.2644      freeview  -4.7972   
4         blair -4.9105             bruvik  -4.2665           wto  -4.8932   
5           baa -5.0375        interactive  -4.4593         rugby  -4.9588   
6        viacom -5.3746              bafta  -4.4676      economic  -5.4020   
7      councils -5.4147             online  -4.9464            g7  -5.7187   
8         straw -6.8016                emi  -5.7958         davos  -5.8680   
9      argonaut -6.8937          gameboyzz  -6.3229         balls  -6.5635   

  sport_word    coef    tech_word     coef  
0  athletics  8.2200          ink  11.2700  
1  liverpool  7.6000     argonaut   9.6500  
2      balco  7.4400         spam   8.9800  
3     doping  7.2800     computer   8.9000  
4      bates  6.8200     software   8.7800  
5     tennis  6.6900       online   8.4300  
6        cup  6.5700    seafarers   8.4100  
7      rugby  6.4500    simonetti   8.3200  
8      conte  6.2400         blog   8.1300  
9       mido  6.1900        robot   7.9700  
0       band -4.1424       shares  -3.6787  
1         cm -4.3488     coverage  -3.6799  
2       film -4.4989        urban  -3.7522  
3   borussia -4.6235  nominations  -3.7691  
4      album -4.6306       plasma  -4.0966  
5   festival -4.9220   complaints  -4.3806  
6     singer -4.9961         corp  -4.9590  
7      piero -5.0658     vodafone  -4.9763  
8    snooker -5.8390      chernin  -6.9950  
9         mr -5.9672          ict  -8.0654